mirrord/layer/src/file/hooks.rs

```diff
@@ -308,12 +308,13 @@ pub(crate) unsafe extern "C" fn openat_detour(
     fd: RawFd,
     raw_path: *const c_char,
     open_flags: c_int,
+    mode: c_int,
 ) -> RawFd {
     let open_options = OpenOptionsInternalExt::from_flags(open_flags);
 
     openat(fd, raw_path.checked_into(), open_options).unwrap_or_bypass_with(|bypass| {
         let raw_path = update_ptr_from_bypass(raw_path, &bypass);
-        FN_OPENAT(fd, raw_path, open_flags)
```

Comment:
Huh, nice catch

Reply:
Ok, there's a problem.

```c
#include <fcntl.h>

int main() {
    char * file_1 = "/tmp/dupa/tmpfile1";
    char * file_2 = "/tmp/dupa/tmpfile2";

    openat(0, file_1, O_CREAT | O_RDWR, 0444);

    openat(0, file_2, O_CREAT | O_RDWR);

    return 0;
}
```

When I run this without mirrord, I get
```
-r--r--r-- 1 razz4780 razz4780 0 Oct 22 11:04 tmpfile1
--wxr----T 1 razz4780 razz4780 0 Oct 22 11:04 tmpfile2
```

When I run this with mirrord, I get different permissions for the second file. Doesn't matter whether I use mode as 4th argument in the detour or if I use vaargs :V

Reply:
I think you need something like
```rs
#[hook_fn]
pub(super) unsafe extern "C" fn open_detour(
    raw_path: *const c_char,
    open_flags: c_int,
    mut args: ...
) -> RawFd {
    let mode: c_int = args.arg();
    let guard = DetourGuard::new();
    if guard.is_none() {
        FN_OPEN(raw_path, open_flags, mode)
    } else {
        open_logic(raw_path, open_flags, mode).unwrap_or_bypass_with(|bypass| {
            let raw_path = update_ptr_from_bypass(raw_path, &bypass);
            FN_OPEN(raw_path, open_flags, mode)
        })
    }
}
```
(from open_detour)

Reply:
Yeah, that's what I did :<
```rust
#[hook_fn]
pub(crate) unsafe extern "C" fn openat_detour(
    fd: RawFd,
    raw_path: *const c_char,
    open_flags: c_int,
    mut args: ...
) -> RawFd {
    let mode: libc::mode_t = args.arg();

    let guard = DetourGuard::new();
    if guard.is_none() {
        FN_OPENAT(fd, raw_path, open_flags, mode)
    } else {
        let open_options = OpenOptionsInternalExt::from_flags(open_flags);

        openat(fd, raw_path.checked_into(), open_options).unwrap_or_bypass_with(|bypass| {
            let raw_path = update_ptr_from_bypass(raw_path, &bypass);
            FN_OPENAT(fd, raw_path, open_flags, mode)
        })
    }
}
```

Reply:
Oh wait.
`mode specifies the permissions to use in case a new file is created. This argument must be supplied when O_CREAT is specified in flags; if O_CREAT is not specified, then mode is ignored.`
(from [man](https://linux.die.net/man/2/open))

I think this means we don't need vaargs at all?